### PR TITLE
Clean up function args

### PR DESCRIPTION
**Pull request by zeroliu**
Clean up constructor and function arguments when singleton can be used. See inline comments for more details

Tests
- QA vault works
- `@vault`, `@web`, `@youtube`, `@pomodoro` works
- removing files index works
- custom db search command works
- updating exclusion path works

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -68,13 +66,12 @@ export default class CopilotPlugin extends Plugin {
     // Always have one instance of sharedState and chainManager in the plugin
     this.sharedState = new SharedState();
 
-    this.vectorStoreManager = new VectorStoreManager(this.app);
-    await this.vectorStoreManager.waitForInitialization();
```
Comment (zeroliu) [inline src/main.ts:72]: no need to explicitly block onload. The initialization will happen in the background and all public methods will wait for initialization.

File: src/search/hybridRetriever.ts [hunk 2]
```diff
@@ -122,11 +108,12 @@ export class HybridRetriever extends BaseRetriever {
...
     try {
       const promptResult = await this.queryRewritePrompt.format({ question: query });
-      const rewrittenQueryObject = await this.llm.invoke(promptResult);
+      const chatModel = ChatModelManager.getInstance().getChatModel();
+      const rewrittenQueryObject = await chatModel.invoke(promptResult);
 
       // Directly return the content assuming it's structured as expected
       if (rewrittenQueryObject && "content" in rewrittenQueryObject) {
-        return rewrittenQueryObject.content;
+        return rewrittenQueryObject.content as string;
```
Comment (zeroliu) [inline src/search/hybridRetriever.ts:116]: there is a type issue here but this refactor doesn't have any logic change. It should be safe to cast to string and keep the same type definition as before.

File: src/search/vectorStoreManager.ts [hunk 3]
```diff
@@ -1,22 +1,22 @@
...
 import { IndexEventHandler } from "./indexEventHandler";
 import { IndexOperations } from "./indexOperations";
 
 export default class VectorStoreManager {
+  private static instance: VectorStoreManager;
   private indexOps: IndexOperations;
   private eventHandler: IndexEventHandler;
   private initializationPromise: Promise<void>;
   private lastKnownSettings: CopilotSettings | undefined;
-  public dbOps: DBOperations;
```
Comment (zeroliu) [inline src/search/vectorStoreManager.ts:16]: dbOps shouldn't be accessed before the initialization finishes

File: src/search/vectorStoreManager.ts [hunk 4]
```diff
@@ -26,11 +26,18 @@ export default class VectorStoreManager {
     this.setupSettingsSubscription();
   }
 
+  static getInstance(): VectorStoreManager {
```
Comment (zeroliu) [inline src/search/vectorStoreManager.ts:29]: made this singleton. It was only instantiated once

File: src/search/vectorStoreManager.ts [hunk 5]
```diff
@@ -52,6 +59,10 @@ export default class VectorStoreManager {
       ) {
         await this.eventHandler.updateExcludedFiles();
       }
+    };
+
+    subscribeToSettingsChange(() => {
+      this.initializationPromise = reinitialize();
```
Comment (zeroliu) [inline src/search/vectorStoreManager.ts:65]: we need to wait for initialization again when settings change causes the singleton to reinitialize

File: src/search/vectorStoreManager.ts [hunk 6]
```diff
@@ -121,15 +132,16 @@ export default class VectorStoreManager {
     this.dbOps.onunload();
   }
 
-  public async getOrInitializeDb(embeddingsAPI: Embeddings): Promise<Orama<any>> {
```
Comment (zeroliu) [inline src/search/vectorStoreManager.ts:124]: We shouldn't have another place to initialize DB. `initialize` method be the one that instantiate db

File: src/search/hybridRetriever.ts [hunk 7]
```diff
@@ -2,28 +2,21 @@ import { BrevilabsClient } from "@/LLMProviders/brevilabsClient";
...
 export class HybridRetriever extends BaseRetriever {
   public lc_namespace = ["hybrid_retriever"];
 
-  private llm: BaseLanguageModel;
   private queryRewritePrompt: ChatPromptTemplate;
-  private embeddingsInstance: Embeddings;
-  private brevilabsClient: BrevilabsClient;
 
   constructor(
-    private dbOps: DBOperations,
```
Comment (zeroliu) [inline src/search/hybridRetriever.ts:22]: hybridRetriever becomes a lot easier to use after taking out these singleton instance args

**Comment by logancyang** [comment]
Nice! Thanks for the cleanup! ✅

**Pull request merged**